### ISSUE TITLE
Can't restore some apps

### ISSUE THREAD

**Issue by lightonflux** [jensstein/oandbackup#77]
I get the following error. It also happens with Wallabad (non-root app).

    AdAway: pkg: /storage/emulated/0/oandbackups/org.adway-2.apk
    AdAway: Failure [INSTALL_FAILED_INVALID_URI] 
    : setPermissions error: could not find permissions for /data/data/org.adaway

The permissions for /data/data are `drwxrwx--x` and the owner is system.

Probably not an oab problem, but i noticed the problem with oab.

Any hints that help me to solve this problem are appreciated.

**Comment by jensstein** [comment]
this could be caused by the bug fixed by this commit: e54f085da400b7012b692fb34c3e5d08127ff72d
could you please check the value for "sourceDir" in the log file for the apps (e.g. oandbackups/org.adaway/org.adaway.log)?
you probably get the error because "/storage/emulated/0/oandbackups/org.adway-2.apk" is not the path to the backed up apk. i would guess it's really "/storage/emulated/0/oandbackups/org.adaway/org.adway-2.apk"

**Comment by lightonflux** [comment]
Vim (Touch version) can't find "sourceDir" in the log. :(

**Comment by jensstein** [comment]
that sounds strange. maybe it has gotten corrupted. could you post the whole log file for adaway (it's probably "/storage/emulated/0/oandbackups/org.adway/org.adaway.log", so this is not oandbackup.log we need (if the terminology is confusing))?

**Comment by lightonflux** [comment]
adaway.log:

        {
            "backupMode": 3,
            "sourceDir": "\/data\/app\/org.adaway-2.apk",
            "packageName": "org.adaway",
            "versionCode": 51,
            "label": "AdAway",
            "dataDir": "\/data\/data\/org.adaway",
            "versionName": "2.9.2",
            "lastBackupMillis": 1419468144580,
            "isSystem": false
        }

I only grepped through the oab.log and didn't found it.

**Comment by jensstein** [comment]
i realise that it's probably not the best choice to call the files with information on the apps logs when there is also the error log.
but the content here is as expected. now, could you please confirm that the backed up apk is indeed called "org.adaway-2.apk"? it might be org.adaway-1.apk and then that would cause a problem.

**Comment by cjeanneret** [comment]
Heya!

I get the same problem with *all* apps when using this app on Slimroms system…

I'll wait for the new release in order to see if it's working better :).

Cheers,

C.

**Comment by jensstein** [comment]
@cjeanneret have you tried the apk i linked to in the other issue?
could you please post the error message and log file (e.g. /sdcard/oandbackups/dk.jens.backup/dk.jens.backup.log) for one of your apps that fail. and check if the value in sourceDir corresponds to an existing file. it might have "/data/app/dk.jens.backup-1.apk" while the actual backed up apk is "/sdcard/oandbackup/dk.jens.backup/dk.jens.backup-2.apk".
i think i'll have to make a method to check for this problem before the next release.
it happens on the current version if a data backup is made while the sourceDir value has changed since the last apk backup.

**Comment by cjeanneret** [comment]
Heya!

I'll try to provide the information this evening, with the new APK and the older.

Cheers,

C.

**Comment by jensstein** [comment]
@cjeanneret is see you wrote in the other issue that the apk works for you. does that mean that you are not seeing this issue any more either?

**Comment by lightonflux** [comment]
Which issue and apk are we talking about? I would like to test it.

**Comment by jensstein** [comment]
it's issue 73. here's the apk: https://drive.google.com/file/d/0B0bNUNWd2wPPVGUtSDhoblhaZzA/view?usp=sharing
please test :) and before you do that it would be really nice if you could compare the value for sourceDir in one of your faulty backups with the filename of the backed up apk.

**Comment by lightonflux** [comment]
So i can start the apk on Android 5.0.2. But it looks like all backups only had data in them, at least the one hi had on my phone at the moment. So i have to check that later.

oab failes to restore data (application is obviously installed).

**Comment by jensstein** [comment]
how does it look like they only have data? does it say so in the app list or are the actual apk files not present in the backup subdirectories?
and please give me the error log file so can see what goes wrong.

**Comment by lightonflux** [comment]
In the UI there is a button for all three options (apt, data, apt+data).  When i press only data restore it silently fails (only visible via the notification). There is no oab.log. And nothing new was written the apps own log. In this case VLC. So i just deleted its log just to be sure, but now oab crashes when i try to restore data of vlc.

> how does it look like they only have data? does it say so in the app list or are the actual apk files not present in the backup subdirectories?

The apks are in the right folders with the right names. But on restore oab displays (when restore data+apk selected):

> com.chmod0.manpages: Cannot restore data without APK, package not installed:

Or the original in german:

> com.chmod0.manpages: Kann Daten nicht ohne APK wiederherstellen, Paket ist nicht installiert:

And there is nothing behind the double point at the end.

**Comment by lightonflux** [comment]
Just realised that the error also appears with fdroid. I can't install applications because the application is not installed. m(  So please ignore this part of the problem.

Probably just a big problem with the ROM.

This should be a problem with oab on all devices.
>  In this case VLC. So i just deleted its log just to be sure, but now oab crashes when i try to restore data of vlc.

**Comment by jensstein** [comment]
yes, you cannot delete the individual logs because they contain data on how the package was when it was backed up. i should make a function to restore them as much as possible sometime.

yes, as you say you probably get the apk error because of some system error. you can try clicking the apk file directly in a file manager or executing `su -c 'pm install -r $path_to_apk'` in a terminal emulator. if you get errors here too then it might be because there are leftover files in the library directories. try uninstalling though oab or try something like `find /data -name $package_name` and delete or move files you find that seem like they might cause problems (and proceed with caution of course).

**Comment by lightonflux** [comment]
So i looked deeper in it. And it looks like oab copies the apk file from the backup with the wrong permissions (0700 instead of 0751) and with the owner as root.  That's why i thought the backup didn't run properly. The restore just moved the apks away. 

But i wasn't able to fix the issue. I just installed all the apps from fdroid and did only a data restore. 

BTW: The restore did restore several apps probably. I have the feeling that maybe the is caused by Firefox. Because both times i tried it with different nightly versions of CM it worked until the batch restore failed at firefox. Firefox is the app with most data 50MB+, maybe that is related. Or maybe it is just a coincidence. 

It works with fresh backups, so i wouldn't mind if we close this issue. But then again it is a backup software, such a problem should not happen (if it is caused by oab rather then the ROM).

**Comment by jensstein** [comment]
if we are talking about user apps and not system apps then oab doesn't copy the apks at all. they are copied to the backup with `cp` but they are installed with the android firmware program `pm` when restoring. so oab has nothing to do with the permissions of a restored user apk (unless something is completely broken, but then the problem would probably have wider consequences).
could you please try installing a failing app manually with `su -c 'pm install -r $path_to_apk'`.

and where do you see the 0700 permissions and root ownship, on the restored apk or on the apk in the backup directory?
if you see it on the backed up file then you might prossibly have your backup directory set to somewhere non-standard. if that's the case, could you please elaborate a little on your setup? ordinarily, files on the storage partition all have something like sd_write as their user and group. and i remember from a previous bug that `pm` cannot install from the data partition, so if the backup is residing in oab's own data directory, the apk is copied to a temporary place on the storage partition and installed from there. it might be this step which fails somehow if your setup is like this.

we should definitely try to determine whether this is a firmware issue or an oab issue. let's keep the issue open until we are sure we cannot fix it (i completely agree that it should not happen if it's oab's fault).
(also, coincidentally i have had problems with updating firefox through f-droid the last couple of versions. it would just stall on the system installer. but maybe it's completely unrelated.)

**Comment by lightonflux** [comment]
> and where do you see the 0700 permissions and root ownship, on the restored apk or on the apk in the backup directory?

/data/data/restored-apk

My oab folder is /storage/emulated/0/oandbackup.

I think that happens when oab restores the data without an apk present. Because i took a look at my Backup drive and there are some apks missing in the backup, which invalidates my theory that the apk was moved not cp'ed. 

But when i try to recreate that with those steps it doesn't work as expected:

1. install app
2. add data
3. backup app with oab
4.  deinstall app
5. delete apk from oab folder
6. restore both

I only get the "can't restore data without apk" message, not the weird "can't install apk without apk" message. 

And i don't get the "apk without apk" error [with fdroid](https://github.com/jensstein/oandbackup/issues/77#issuecomment-71322140). In other words i don't get the error anywhere.

So it probably is the ROM, because fdroid didn't change, only the rom. And i also applied [this](https://mobiledevstories.wordpress.com/2013/06/03/android-adb-install_failed_invalid_uri/), just in case. 



> su -c 'pm install -r $path_to_apk'.

Works with absolute path. But if i use a relative path i get the invalid URI message (already in a root shell, no su, so the pwd is right). I guess that is expected behavior for the tool?!


In conclusion: I think the problem was my incomplete backup (without some apks) in conjugation with a permission error.

**Comment by jensstein** [comment]
i think you're right that the problem may be two separate problems at the same time. because oab should have nothing to do with apk permissions. the missing apks are probably due to the bug fixed here: e228dd2687d3923bc3e62d3849c1818daab59d9c
oab always uses `pm install` with absolute paths so that shouldn't be a problem.
let's close the issue, then you can just re-open if you get the issue again. but i hope it's fixed for you now :)

(also, i'm sure you know but you shouldn't ever have to set the permissions as 777 on anything like they do in your last link. the permissions of my /data/local are 755 and that works fine.)
